src/testing.py

```diff
@@ -89,6 +89,46 @@ def __call__(self, test_file):
       return Result(test=basename, success=False, output=e.output)
 
 
+def parse_exclude_files(fails, config_attributes):
+  ''' Parse the files containing tests to exclude (i.e. expected fails).
+  Each line may contain a comma-separated list of attributes restricting
+  the test configurations which are expected to fail. (e.g. JS engine
+  or optimization level). A test is only excluded if the configuration
+  has all the attributes specified in the exclude line. Lines which
+  have no attributes will match everything, and lines which specify only
+  one attribute (e.g. engine) will match all configurations with that
+  attribute (e.g. both opt levels with that engine). Returns a sorted list
+  of exclusions which match the attributes.
+  '''
+  excludes = {}  # maps name of excluded test to file from whence it came
+  if not config_attributes:
+    config_attributes = set()
+  for excludefile in fails:
+    f = open(excludefile)
+    for line in f:
+      line = line.strip()
+      if '#' in line:
+        line = line[:line.index('#')].strip()
+      if not line:
+        continue
+      tokens = line.split()
+      if len(tokens) > 1:
+        attributes = set(tokens[1].split(','))
+        if not attributes.issubset(config_attributes):
+          continue
+        test = tokens[0]
+      else:
+        test = line
+      if test in excludes:
+        print 'ERROR: duplicate exclude: [%s]' % line
+        print 'Files: %s and %s' % (excludes[test], excludefile)
+        sys.exit(1)
```

Comment:
Is this a hard error?
What if we wanted something like
```
# foo.c only works with -O0
foo.c o1
foo.c o2
```

Should we add a magic `known-dupe` attribute? (no we should not)
Should it be a warning? (we'll still want the attribute to silence the warnings I'll bet)
Should we not care? (we'll have duplicates and not know)
I vote we don't care, because when the status changes and we go to remove one we'll see it still isn't running.
Turned around, what's the reason we want to disallow duplicate excludes?